### ISSUE TITLE
False positive when using negated branch: `possibly-used-before-assignment`

### ISSUE THREAD

**Issue by Flamefire** [pylint-dev/pylint#10804]
### Bug description

```python
def x(a):
    if a:
        name = 'A'
    if not a:
        print(1)
    else:
        print(name) # possibly-used-before-assignment
    if a:
        print(name) # OK
    else:
        print(1)
```

### Configuration

```ini

```

### Command used

```shell
pylint /tmp/test.py
```

### Pylint output

```python
/tmp/test.py:7:14: E0606: Possibly using variable 'name' before assignment (possibly-used-before-assignment)
```

### Expected behavior

No error.

Note that order matters: Placing the `if a` part on top makes the error disappear.
When using only one variant, then only the `if not a - else` fails

### Pylint version

```shell
pylint 4.1.0-dev0
astroid 4.0.2
Python 3.10.13 (main, Sep 17 2024, 16:25:46) [GCC 11.4.0]
```

**Comment by jacobtylerwalls** [comment]
Thanks for the report. This is the same class of report as #9662. We can't start working on these issues until we start checking for anything that might have happened between `if a:` and `if not a:`, since any reassignment to `a` in between will throw off the analysis.

**Comment by Flamefire** [comment]
Yes looks similar.

> since any reassignment to a in between will throw off the analysis.

I'd expect that this can be detected from the AST and if no assignment happens the value could be (re)used, can't it? That would make it a bit better than what seems to be the current state of: when "if a" == "if a" consider as taken

I can see how a general solution is hard. I.e. `if a` and `if not a` might be easy to infer but an `and` or `or` will be much harder

**Comment by jacobtylerwalls** [comment]
Yep, I think that's the plan, just needs a volunteer.
